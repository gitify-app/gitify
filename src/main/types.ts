export interface GitifyAPI {
  openExternalLink: (url: string) => void;
  getAppVersion: () => Promise<string>;
  encryptValue: (value: string) => Promise<string>;
  decryptValue: (value: string) => Promise<string>;
  quitApp: () => void;
  showWindow: () => void;
  hideWindow: () => void;
  setAutoLaunch: (value: boolean) => void;
  setAlternateIdleIcon: (value: boolean) => void;
  setKeyboardShortcut: (keyboardShortcut: boolean) => void;
  updateTrayIcon: (notificationsLength?: number) => void;
  updateTrayTitle: (title?: string) => void;
}
