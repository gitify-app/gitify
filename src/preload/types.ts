import type { api } from '.';

export type GitifyAPI = typeof api;
