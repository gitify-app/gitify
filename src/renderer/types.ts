import type { OcticonProps } from '@primer/octicons-react';
import type { FC } from 'react';
import type { Notification, Reason } from './typesGitHub';
import type {
  AuthMethod,
  EnterpriseAccount,
  PlatformType,
} from './utils/auth/types';

declare const __brand: unique symbol;

type Brand<B> = { [__brand]: B };

export interface AuthState {
  accounts: Account[];
  /**
   * @deprecated This attribute is deprecated and will be removed in a future release.
   */
  token?: Token;
  /**
   * @deprecated This attribute is deprecated and will be removed in a future release.
   */
  enterpriseAccounts?: EnterpriseAccount[];
  /**
   * @deprecated This attribute is deprecated and will be removed in a future release.
   */
  user?: GitifyUser | null;
}

export type Branded<T, B> = T & Brand<B>;

export type AuthCode = Branded<string, 'AuthCode'>;

export type Token = Branded<string, 'Token'>;

export type ClientID = Branded<string, 'ClientID'>;

export type ClientSecret = Branded<string, 'ClientSecret'>;

export type Hostname = Branded<string, 'Hostname'>;

export type Link = Branded<string, 'WebUrl'>;

export type Status = 'loading' | 'success' | 'error';

export interface Account {
  method: AuthMethod;
  platform: PlatformType;
  version?: string;
  hostname: Hostname;
  token: Token;
  user: GitifyUser | null;
  hasRequiredScopes?: boolean;
}

export type SettingsValue =
  | boolean
  | number
  | GroupBy
  | OpenPreference
  | Reason[]
  | Theme;

export type SettingsState = AppearanceSettingsState &
  NotificationSettingsState &
  SystemSettingsState &
  FilterSettingsState;

export interface AppearanceSettingsState {
  theme: Theme;
  zoomPercentage: number;
  detailedNotifications: boolean;
  showAccountHeader: boolean;
  showPills: boolean;
  showNumber: boolean;
  wrapNotificationTitle: boolean;
}

export interface NotificationSettingsState {
  groupBy: GroupBy;
  fetchAllNotifications: boolean;
  participating: boolean;
  markAsDoneOnOpen: boolean;
  markAsDoneOnUnsubscribe: boolean;
  delayNotificationState: boolean;
}

export interface SystemSettingsState {
  openLinks: OpenPreference;
  keyboardShortcut: boolean;
  showNotificationsCountInTray: boolean;
  showNotifications: boolean;
  useAlternateIdleIcon: boolean;
  playSound: boolean;
  openAtStartup: boolean;
}

export interface FilterSettingsState {
  hideBots: boolean;
  filterReasons: Reason[];
}

export interface GitifyState {
  auth?: AuthState;
  settings?: SettingsState;
}

export enum Theme {
  SYSTEM = 'SYSTEM',
  LIGHT = 'LIGHT',
  LIGHT_HIGH_CONTRAST = 'LIGHT_HIGH_CONTRAST',
  LIGHT_COLORBLIND = 'LIGHT_COLORBLIND',
  LIGHT_TRITANOPIA = 'LIGHT_TRITANOPIA',
  DARK = 'DARK',
  DARK_HIGH_CONTRAST = 'DARK_HIGH_CONTRAST',
  DARK_COLORBLIND = 'DARK_COLORBLIND',
  DARK_TRITANOPIA = 'DARK_TRITANOPIA',
  DARK_DIMMED = 'DARK_DIMMED',
}

export enum OpenPreference {
  FOREGROUND = 'FOREGROUND',
  BACKGROUND = 'BACKGROUND',
}

export enum GroupBy {
  REPOSITORY = 'REPOSITORY',
  DATE = 'DATE',
}

export type RadioGroupItem = {
  label: string;
  value: string;
};

export interface AccountNotifications {
  account: Account;
  notifications: Notification[];
  error: GitifyError | null;
}

export interface GitifyUser {
  login: string;
  name: string | null;
  avatar: Link | null;
  id: number;
}

export interface GitifyError {
  title: string;
  descriptions: string[];
  emojis: string[];
}

export type ErrorType =
  | 'BAD_CREDENTIALS'
  | 'MISSING_SCOPES'
  | 'NETWORK'
  | 'RATE_LIMITED'
  | 'UNKNOWN';

export interface FormattedReason {
  title: string;
  description: string;
}

export enum IconColor {
  GRAY = 'text-gitify-icon-muted',
  GREEN = 'text-gitify-icon-open',
  PURPLE = 'text-gitify-icon-done',
  RED = 'text-gitify-icon-closed',
  YELLOW = 'text-gitify-icon-attention',
}

export enum Opacity {
  READ = 'opacity-50',
  LOW = 'opacity-70',
  MEDIUM = 'opacity-80',
  HIGH = 'opacity-90',
}

export type PullRequestApprovalIcon = {
  type: FC<OcticonProps>;
  color: IconColor;
  description: string;
};

export enum Size {
  XSMALL = 12,
  SMALL = 14,
  MEDIUM = 16,
  LARGE = 18,
  XLARGE = 20,
}

export type Chevron = {
  icon: FC<OcticonProps>;
  label: string;
};
