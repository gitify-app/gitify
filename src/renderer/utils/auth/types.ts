import type {
  AuthCode,
  ClientID,
  ClientSecret,
  Hostname,
  Token,
} from '../../types';

export type AuthMethod = 'GitHub App' | 'Personal Access Token' | 'OAuth App';

export type PlatformType = 'GitHub Cloud' | 'GitHub Enterprise Server';

export interface LoginOAuthAppOptions {
  hostname: Hostname;
  clientId: ClientID;
  clientSecret: ClientSecret;
}

export interface LoginPersonalAccessTokenOptions {
  hostname: Hostname;
  token: Token;
}

export interface AuthResponse {
  authMethod: AuthMethod;
  authCode: AuthCode;
  authOptions: LoginOAuthAppOptions;
}

export interface AuthTokenResponse {
  hostname: Hostname;
  token: Token;
}
