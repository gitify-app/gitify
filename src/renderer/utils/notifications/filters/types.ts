import type {
  AccountNotifications,
  SettingsState,
  TypeDetails,
} from '../../../types';
import type { Notification } from '../../../typesGitHub';

export interface Filter<T extends string> {
  FILTER_TYPES: Record<T, TypeDetails>;

  requiresDetailsNotifications: boolean;

  getTypeDetails(type: T): TypeDetails;

  hasFilters(settings: SettingsState): boolean;

  isFilterSet(settings: SettingsState, type: T): boolean;

  getFilterCount(notifications: AccountNotifications[], type: T): number;

  filterNotification(notification: Notification, type: T): boolean;
}
