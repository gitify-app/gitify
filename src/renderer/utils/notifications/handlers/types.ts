import type { FC } from 'react';

import type { OcticonProps } from '@primer/octicons-react';

import type { SettingsState } from '../../../types';
import type {
  GitifySubject,
  Notification,
  Subject,
  SubjectType,
} from '../../../typesGitHub';

export interface NotificationTypeHandler {
  readonly type?: SubjectType;

  /**
   * Enrich a notification. Settings may be unused for some handlers.
   */
  enrich(
    notification: Notification,
    settings: SettingsState,
  ): Promise<GitifySubject>;

  /** Return an icon component for this notification type. */
  getIcon(subject: Subject): FC<OcticonProps> | null;
}
