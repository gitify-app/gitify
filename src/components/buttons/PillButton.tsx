import type { Icon } from '@primer/octicons-react';
import type { FC } from 'react';
import type { IconColor } from '../../types';

export interface IPillButton {
  key?: string;
  title: string;
  metric?: number;
  icon: Icon;
  color: IconColor;
}

export const PillButton: FC<IPillButton> = (props: IPillButton) => {
  return (
    <span title={props.title}>
      <button
        type="button"
        className="rounded-full text-xss px-1 m-0.5 bg-gray-100 hover:bg-gray-200 dark:bg-gray-800 dark:hover:bg-gray-700"
      >
        <props.icon
          size={12}
          className={`mr-1 ${props.color}`}
          aria-label={props.title}
        />
        {props.metric}
      </button>
    </span>
  );
};
