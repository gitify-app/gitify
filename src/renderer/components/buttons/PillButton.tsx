import type { FC } from 'react';

import type { Icon } from '@primer/octicons-react';
import { Octicon, Stack, Text } from '@primer/react';

import { type IconColor, Size } from '../../types';

export interface IPillButton {
  key?: string;
  title: string;
  metric?: number;
  icon: Icon;
  color: IconColor;
}

export const PillButton: FC<IPillButton> = (props: IPillButton) => {
  return (
    <button
      title={props.title}
      type="button"
      className="flex gap-1 items-center text-xxs px-1 m-0.5 rounded-full bg-gitify-pill-rest hover:bg-gitify-pill-hover"
    >
      <Stack direction="horizontal" align="center" gap="condensed">
        <Octicon icon={props.icon} size={Size.XSMALL} className={props.color} />
        {props.metric && <Text>{props.metric}</Text>}
      </Stack>
    </button>
  );
};
