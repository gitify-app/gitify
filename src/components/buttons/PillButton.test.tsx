import { MarkGithubIcon } from '@primer/octicons-react';
import { render } from '@testing-library/react';
import { IconColor } from '../../types';
import { type IPillButton, PillButton } from './PillButton';

describe('components/buttons/PillButton.tsx', () => {
  it('should render', () => {
    const props: IPillButton = {
      title: 'Mock Pill',
      metric: 1,
      icon: MarkGithubIcon,
      color: IconColor.GREEN,
    };
    const tree = render(<PillButton {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
