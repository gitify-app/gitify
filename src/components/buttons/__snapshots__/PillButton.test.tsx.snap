// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/buttons/PillButton.tsx should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        title="Mock Pill"
      >
        <button
          class="m-0.5 rounded-full bg-gray-100 px-1 text-xss hover:bg-gray-200 dark:bg-gray-800 dark:hover:bg-gray-700"
          type="button"
        >
          <svg
            aria-label="Mock Pill"
            class="mr-1 text-green-500"
            fill="currentColor"
            focusable="false"
            height="12"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 16 16"
            width="12"
          >
            <path
              d="M8 0c4.42 0 8 3.58 8 8a8.013 8.013 0 0 1-5.45 7.59c-.4.08-.55-.17-.55-.38 0-.27.01-1.13.01-2.2 0-.75-.25-1.23-.54-1.48 1.78-.2 3.65-.88 3.65-3.95 0-.88-.31-1.59-.82-2.15.08-.2.36-1.02-.08-2.12 0 0-.67-.22-2.2.82-.64-.18-1.32-.27-2-.27-.68 0-1.36.09-2 .27-1.53-1.03-2.2-.82-2.2-.82-.44 1.1-.16 1.92-.08 2.12-.51.56-.82 1.28-.82 2.15 0 3.06 1.86 3.75 3.64 3.95-.23.2-.44.55-.51 1.07-.46.21-1.61.55-2.33-.66-.15-.24-.6-.83-1.23-.82-.67.01-.27.38.01.53.34.19.73.9.82 1.13.16.45.68 1.31 2.69.94 0 .67.01 1.3.01 1.49 0 .21-.15.45-.55.38A7.995 7.995 0 0 1 0 8c0-4.42 3.58-8 8-8Z"
            />
          </svg>
          1
        </button>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      title="Mock Pill"
    >
      <button
        class="m-0.5 rounded-full bg-gray-100 px-1 text-xss hover:bg-gray-200 dark:bg-gray-800 dark:hover:bg-gray-700"
        type="button"
      >
        <svg
          aria-label="Mock Pill"
          class="mr-1 text-green-500"
          fill="currentColor"
          focusable="false"
          height="12"
          role="img"
          style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
          viewBox="0 0 16 16"
          width="12"
        >
          <path
            d="M8 0c4.42 0 8 3.58 8 8a8.013 8.013 0 0 1-5.45 7.59c-.4.08-.55-.17-.55-.38 0-.27.01-1.13.01-2.2 0-.75-.25-1.23-.54-1.48 1.78-.2 3.65-.88 3.65-3.95 0-.88-.31-1.59-.82-2.15.08-.2.36-1.02-.08-2.12 0 0-.67-.22-2.2.82-.64-.18-1.32-.27-2-.27-.68 0-1.36.09-2 .27-1.53-1.03-2.2-.82-2.2-.82-.44 1.1-.16 1.92-.08 2.12-.51.56-.82 1.28-.82 2.15 0 3.06 1.86 3.75 3.64 3.95-.23.2-.44.55-.51 1.07-.46.21-1.61.55-2.33-.66-.15-.24-.6-.83-1.23-.82-.67.01-.27.38.01.53.34.19.73.9.82 1.13.16.45.68 1.31 2.69.94 0 .67.01 1.3.01 1.49 0 .21-.15.45-.55.38A7.995 7.995 0 0 1 0 8c0-4.42 3.58-8 8-8Z"
          />
        </svg>
        1
      </button>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
